contracts/src/Teleporter/TeleporterMessenger.sol

```diff
@@ -377,19 +377,37 @@ contract TeleporterMessenger is ITeleporterMessenger, ReentrancyGuards {
         emit MessageExecuted(originChainID, message.messageID);
         delete receivedFailedMessageHashes[originChainID][message.messageID];
 
+        // Re-encode the payload by ABI encoding a call to the {receiveTeleporterMessage} function
+        // defined by the {ITeleporterReceiver} interface.
+        bytes memory payload = abi.encodeCall(
+            ITeleporterReceiver.receiveTeleporterMessage,
+            (originChainID, message.senderAddress, message.message)
+        );
+
         // Reattempt the message execution with all of the gas left available for execution of this transaction.
         // Use all of the gas left because this message has already been successfully delivered, and it is the
         // responsibility of the caller to provide as much gas is needed. Compared to the initial delivery, where
         // the relayer should still receive their reward even if the message execution takes more gas than expected.
         // Require that the call be successful such that in the failure case this transaction reverts and the
         // message can be retried again if desired.
-        // solhint-disable-next-line avoid-low-level-calls
-        (bool success, ) = message.destinationAddress.call(
-            abi.encodeCall(
-                ITeleporterReceiver.receiveTeleporterMessage,
-                (originChainID, message.senderAddress, message.message)
+        //
+        // Use assembly to make the low-level call to avoid unnecessary memory expansion of the return data
+        // due to an issue with the version of Solidity used. This prevents possible "return bomb" vectors
+        // where the external contract could for the caller to use an arbitrary amount of gas.
+        address target = message.destinationAddress;
+        bool success;
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            success := call(
+                gas(), // gas provided to the call
+                target, // call target
+                0, // zero value
+                add(payload, 0x20), // input data
+                mload(payload), // input data length
```

Comment:
From the Yul [documentation](https://docs.soliditylang.org/en/latest/yul.html#evm-dialect), these arguments correspond to the input and input size. Can you explain how the `add` and `mload` operations correspond to this? I'd expect to be something like `mload(payload)` to get the input pointer, then `msize(payload)` to get the payload size.

Reply:
To be honest, I pulled this from a handful of examples I found online and from OpenZeppelin. Dug into it a bit more now, and added comments on the usage of `add` and `mload`. Thanks for calling out.

Reply:
Thanks, yeah those comments clarify things. Do you know if the array length is always guaranteed to be in the first slot of the array? How about for an empty array?

Reply:
From the [Solidity docs](https://docs.soliditylang.org/en/v0.8.17/internals/layout_in_storage.html#bytes-and-string):

"""
bytes and string are encoded identically. In general, the encoding is similar to bytes1[], in the sense that there is a slot for the array itself and a data area that is computed using a keccak256 hash of that slot’s position. However, for short values (shorter than 32 bytes) the array elements are stored together with the length in the same slot.

In particular: if the data is at most 31 bytes long, the elements are stored in the higher-order bytes (left aligned) and the lowest-order byte stores the value length * 2. For byte arrays that store data which is 32 or more bytes long, the main slot p stores length * 2 + 1 and the data is stored as usual in keccak256(p). This means that you can distinguish a short array from a long array by checking if the lowest bit is set: short (not set) and long (set).
"""

From this, it seems like we need to handle small and large payloads differently: small payloads encode the length and payload bytes in the same slot, where large payloads put the payload bytes at the keccack hash of the length pointer.

Reply:
I think the docs you link refer to the layout of variables in storage, not in memory. I'm checking to see if I can find the correct documentation for encoding of arrays in memory.